src/DocumentConstructionObserver.js

```diff
@@ -0,0 +1,61 @@
+import CustomElementInternals from './CustomElementInternals';
+
+export default class DocumentConstructionObserver {
+  constructor(internals, doc) {
+    /**
+     * @type {!CustomElementInternals}
+     */
+    this._internals = internals;
+
+    /**
+     * @type {!Document}
+     */
+    this._document = doc;
+
+    /**
+     * @type {MutationObserver|undefined}
+     */
+    this._observer = undefined;
+
+
+    // Simulate tree construction for all currently accessible nodes in the
+    // document.
+    this._internals.patchTree(this._document);
+    this._internals.upgradeTree(this._document);
+
+    if (this._document.readyState === 'loading') {
+      this._observer = new MutationObserver(this._handleMutations.bind(this));
+
+      // Nodes created by the parser are given to the observer *before* the next
+      // task runs. Inline scripts are run in a new task. This means that the
+      // observer will be able to handle the newly parsed nodes before the inline
+      // script is run.
+      this._observer.observe(this._document, {
+        childList: true,
+        subtree: true,
+      });
+    }
+  }
+
+  /**
+   * @param {!Array<!MutationRecord>} mutations
+   */
+  _handleMutations(mutations) {
+    // Once the document's `readyState` is 'interactive' or 'complete', all new
+    // nodes created within that document will be the result of script and
+    // should be handled by patching.
+    const readyState = this._document.readyState;
+    if (readyState === 'interactive' || readyState === 'complete') {
```

Comment:
Why is the disconnection done here (in `_handleMutations`) and not in a "readystatechange` event handler?  Seems like this would double-walk the first script-created elements appended to the document after 'interactive'...?